File: c++/cote/GJM/변화하는_중간값.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <fstream>
#include <vector>
#include <queue>

using namespace std;

struct RNG {
	int seed, a, b;
	RNG(int a, int b) : a(a), b(b), seed(1983) {}
	
	int next() {
		int ret = seed;
		seed = ((seed * (long long) a) + b) % 20090711;

		return ret;
	}
};

int* split(const string& s) {
	stringstream ss(s);
	string buf;
	int* ret = new int[3];

	for(int i =0; i< 3; i++){
		getline(ss, buf, ' ');
		ret[i] = stoi(buf);
	}

	return ret;
}

int solve(int n, RNG rng){
	priority_queue<int, vector<int>, less<int>> maxHeap;
	priority_queue<int, vector<int>, greater<int>> minHeap;
	int ret = 0;

	/*
	반복문 불변식 :
	1. maxHeap의 크기는 minHeap의 크기와 같거나 1 더 크다 
	2. maxHeap.top() <= minHeap.top()
	*/

	for (int i=0; i<n; i++) {
		if (maxHeap.size() == minHeap.size()) {
			maxHeap.push(rng.next());
		} else {
			minHeap.push(rng.next());
		}

		// 2번 불변식이 깨질 경우 복구한다 
		if (!minHeap.empty() && !maxHeap.empty() && minHeap.top() < maxHeap.top()){

			int a = maxHeap.top();
			int b = minHeap.top();

			maxHeap.pop(); minHeap.pop();
			
			maxHeap.push(b); 
			minHeap.push(a);
		}

		ret = (ret + maxHeap.top()) % 20090711;
	}

	return ret;
}

int main(void){
	string buf;
	fstream fd ("data/변화하는_중간값.txt");
	
	getline(fd, buf);

	int testCase = stoi(buf);

	while (testCase > 0) {

		getline(fd, buf);
		int* nums = split(buf);
		
		getline(fd, buf);
		int answer = stoi(buf);

		RNG rng(nums[1], nums[2]);

		cout << "answer : " << answer << ", solved : " << solve(nums[0], rng) <<endl;
		testCase--;
	}

	return 0;
}

File: go/cote/GJM/변화하는_중간값.go
package main

import (
	"container/heap"
	"fmt"
	"math/rand"
)

const (
	TEST_CASES = 10
	N          = 10000
	MAX_VALUE  = 10000
)

type Heap struct {
	Data     []int
	LessFunc func(i, j int) bool
}

func (h Heap) Len() int {
	return len(h.Data)
}

func (h Heap) Swap(i, j int) {
	h.Data[i], h.Data[j] = h.Data[j], h.Data[i]
}

func (h Heap) Less(i, j int) bool {
	return h.LessFunc(h.Data[i], h.Data[j])
}

func (h *Heap) Push(x any) {
	h.Data = append(h.Data, x.(int))
}

func (h *Heap) Pop() any {
	old := h.Data
	n := len(old)
	x := old[n-1]
	h.Data = old[0 : n-1]
	return x
}

func minmax(x, y int) (min, max int) {
	if x > y {
		min, max = y, x
		return
	}
	min, max = x, y
	return
}

func solve(nums []int) []int {
	ret := make([]int, 0)

	minheap := &Heap{[]int{}, func(i, j int) bool { return i < j }}
	maxheap := &Heap{[]int{}, func(i, j int) bool { return i > j }}

	/*
		1. minHeap의 크기는 maxHeap의 크기와 같거나 크다
		2. maxHeap의 top <= minHeap의 top
	*/

	for _, v := range nums {
		if minheap.Len() == maxheap.Len() {
			heap.Push(minheap, v)
		} else {
			heap.Push(maxheap, v)
		}

		if minheap.Len() != 0 && maxheap.Len() != 0 {
			x := heap.Pop(minheap)
			y := heap.Pop(maxheap)

			min, max := minmax(x.(int), y.(int))
			heap.Push(maxheap, min)
			heap.Push(minheap, max)
		}

		x := heap.Pop(minheap)
		ret = append(ret, x.(int))
		heap.Push(minheap, x)
	}

	return ret
}

func main() {
	random := rand.New(rand.NewSource(43))

	for it := 0; it < TEST_CASES; it++ {
		n := random.Intn(N-3) + 3
		nums := make([]int, n)

		for i, _ := range nums {
			nums[i] = random.Intn(MAX_VALUE)
		}

		fmt.Println(solve(nums))
	}
}
